stdlib/public/Concurrency/Task.cpp

```diff
@@ -1030,7 +1030,12 @@ swift_task_addCancellationHandlerImpl(
   auto *record = new (allocation)
       CancellationNotificationStatusRecord(unsigned_handler, context);
 
-  swift_task_addStatusRecord(record);
+  if (swift_task_addStatusRecord(record))
+    return record;
+
+  // else, the task was already cancelled, so while the record was added,
+  // we must run it immediately here since no other task will trigger it.
+  record->run();
```

Comment:
The change™.

We used to do this in a racy way in Swift but have to do so in here, based on the return of addStatusRecord